plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -204,14 +187,35 @@ InputStream readBlob(String blobName) throws IOException {
      * @param blobSize    expected size of the blob to be written
      */
     void writeBlob(String blobName, InputStream inputStream, long blobSize) throws IOException {
-        SocketAccess.doPrivilegedVoidIOException(() -> {
-            InputStreamContent stream = new InputStreamContent(null, inputStream);
-            stream.setLength(blobSize);
+        final BlobInfo blobInfo = BlobInfo.newBuilder(bucket, blobName).build();
+        if (blobSize > (5 * 1024 * 1024)) {
+            // uses "resumable upload" for files larger than 5MB, see
+            // https://cloud.google.com/storage/docs/json_api/v1/how-tos/multipart-upload
+            final WriteChannel writeChannel = SocketAccess.doPrivilegedIOException(() -> storage.writer(blobInfo));
+            Streams.copy(inputStream, java.nio.channels.Channels.newOutputStream(new WritableByteChannel() {
+                @Override
+                public boolean isOpen() {
+                    return writeChannel.isOpen();
+                }
 
-            Storage.Objects.Insert insert = client.objects().insert(bucket, null, stream);
-            insert.setName(blobName);
-            insert.execute();
-        });
+                @Override
+                public void close() throws IOException {
+                    SocketAccess.doPrivilegedVoidIOException(writeChannel::close);
+                }
+
+                @SuppressForbidden(reason = "Channel is based of a socket not a file.")
+                @Override
+                public int write(ByteBuffer src) throws IOException {
+                    return SocketAccess.doPrivilegedIOException(() -> writeChannel.write(src));
+                }
+            }));
+        } else {
+            // uses multipart upload for small files (1 request for both data and metadata,
+            // gziped)
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream(Math.toIntExact(blobSize));
+            Streams.copy(inputStream, baos);
+            SocketAccess.doPrivilegedVoidIOException(() -> storage.create(blobInfo, baos.toByteArray()));
+        }
```

Comment:
This is a legit optimization according to the docs, but actually it is necessary to get around the library, which [does not use the port](https://github.com/GoogleCloudPlatform/google-cloud-java/blob/52b727aef88ae76984aa3c02b4d7067e198d34b7/google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java#L720) for the new connections it opens for the "resumable" upload.
This means the REST Integ tests will try to upload files using the default port 80, but the fixture cannot bind this port.
The multipart upload builds the URL correctly.